### DISCUSSION TITLE
Idea for WiFi configuration.

### DISCUSSION THREAD

**Discussion by minecraftchest1** [PlummersSoftwareLLC/NightDriverStrip#5] [Ideas]
I just finished the video and noticed that you talked about wifi manager. While I don't know how that works, would it be possible to start a hotspot and have the webserver setup the wifi connection based on user input from the web interface?

You would want to have the server cache the credintals to a file though.

**Comment by davepl** [comment]
Yes, and it caches them to NVRAM.  The code is there but commented out so hopefully someone with some previous experience with that module can take a look!

**Reply by minecraftchest1** [reply]
That was fast. I was about to look at the code to see.

**Reply by minecraftchest1** [reply]
Well darn. I may have to wait for your next video to figure out how the web code works.  Having a hard time getting the tabs to work. Only the first tab renders anything. Time to dig in the code some more.

**Comment by davepl** [comment]
If you mean the WebServer that runs on the chip, there is only a very basic skeletal page there with an empty Help tab and an effects selector.  The rest is TBD!

**Reply by minecraftchest1** [reply]
I got a fork at https://github.com/minecraftchest1/NightDriverStrip that I added a test tab to. I can uncomment the buttons in tab-1 and added a tab-3, but only tab1 shows anything.
![image](https://user-images.githubusercontent.com/42992675/135760935-684aa868-91b4-459e-9828-c13145696e2b.png)
![image](https://user-images.githubusercontent.com/42992675/135760943-3329431f-e20a-475b-940a-1bbcafb9bd96.png)
![image](https://user-images.githubusercontent.com/42992675/135760954-f31bfa67-9183-461c-b425-4919292202b8.png)
![image](https://user-images.githubusercontent.com/42992675/135760966-ece07b97-82f1-45bd-9893-a2615dc0b0e2.png)

While I doubt this would make a difference, I am just serving the webpage locally using apache, rather than running it on an esp32 (as I do not currently have one).

Surprisingly, if I replace the links with buttons, it renders everything.
![image](https://user-images.githubusercontent.com/42992675/135761084-e0659842-d81a-4822-a167-9ca70c067556.png)
![image](https://user-images.githubusercontent.com/42992675/135761099-0834b4f9-ddb8-49a4-9108-8230bf6f4ad8.png)

**Comment by davepl** [comment]
That won't work, I don't think.  The web page needs to be served by the ESP32 so that the REST API that fetches the current effects and the effects list will be called.  Look at the SPECTRUM project.  You can build that, connected to a simple strip if you don't have a matrix, and upload it.  Then once the code is uploaded, in PlatformIO do a "Builds Filesystem Image SPIFFS" and then an "Upload Filesystem Image" and it should upload the files you need.  Set the WiFi credentials so it can connect to your WiFi, and then watch the serial port (or your router) to figure out what IP is gets.  Browse to that page and you should get a UI like this.

The example I just pasted here is live off a SPECTRUM running here locally, so "it works in my office!" as they say!

![2021-10-03_16-45-00](https://user-images.githubusercontent.com/3698967/135775885-1bf201cd-3ebb-44e4-8207-fb89a2c9292e.png)

**Comment by bst00** [comment]
Great project and great videos Dave, yet another rabbit hole for me to fall into.  I now have a single strip of 144 LEDs running the Demo, more to follow I'm sure.

Have you considered using SmartConfig rather than WiFiManager?  I've only recently started to play with ESP32 and ESP32-CAM and like SmartConfig (I've no experience with WiFiManager yet) as there is no need to embed your ssid or password in any files.  No need to look at the serial monitor to find the IP address assigned to the ESP either, as the app on your smartphone will show that once successfully configured.

**Reply by robertlipe** [reply]
IMO, a SmartConfig implementation would be awesome. It would marry it MORE to ESP32 and LESS to Arduino, which can be argued to be either good or bad, but the feature set is awesome.

The SmartConfig API is dead simple: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_smartconfig.html

Picking through Unified Provisioning, SmartConfig, EasyConnect is a bit of an exercise and it'd be interest to see if civilian users consider installing an app on their fones to be better or worse.

Have you actually done this, especially at a scale where you've had the feedback of a few hundred users? I don't think that actually implementing any one of these approaches (and it seems like there are more) is hard , but figuring out WHICH approach is likely to work for the MAJORITY of users is a project of its own.

As clammy as I am to the implementation of ImprovSerial, I'll readily admit that typing words into a web brower's form page is an interface that 100% of this project's audience is likely to be comfortable with. Installing another app on a fone might not go over so well. Since MOST of our configurations don't have screens, QR methods don't really work.
If anything, ESP-IDF offers TOO MANY good options here. The Bluetooth options may not work well as we support chips that don't support the legacy BT protocols. 

Do you have an implementation in hand? Market research needed/welcome. Will people install an app or leave your project immediately? Do people get confused by SoftAP vs. SmartConfig and bail? IMO, we definitely do NOT want to require a custom mobile build of a configurator just because I"m pretty sure that we don't want to enter the mobile app biz.

If you have a PR that demonstrates replacing Improv with SmartConfig, I'd really like to see it.
